### PR TITLE
CLN: add typing to dtype arg in selection of files in core/arrays (GH38808)

### PR DESCRIPTION
**Pull request by avinashpancham**
Follow on PR for #38808
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/arrays/boolean.py [hunk 1]
```diff
@@ -273,7 +273,7 @@ def dtype(self) -> BooleanDtype:
 
     @classmethod
     def _from_sequence(
-        cls, scalars, *, dtype=None, copy: bool = False
+        cls, scalars, *, dtype: Optional[DtypeArg] = None, copy: bool = False
```
Comment (jreback) [inline pandas/core/arrays/boolean.py:276]: most of these are Optional[Dtype} NOT DtypeArg (eg. only a single dtype and not a dict of them). The only things that take DtypeArg are things like astype and the I/O methods

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/boolean.py (hunk 1 above)
Comment (avinashpancham) [reply pandas/core/arrays/boolean.py]: Changed it. I did it right last time, but I think I was a bit too quick with submitting this PR. Thanks for the review.

**Comment by avinashpancham** [comment]
Some checks are failing, will look at it tomorrow.

File: pandas/core/arrays/categorical.py [hunk 2]
```diff
@@ -1294,7 +1297,7 @@ def _validate_fill_value(self, fill_value):
 
     # -------------------------------------------------------------
 
-    def __array__(self, dtype=None) -> np.ndarray:
+    def __array__(self, dtype: Optional[Dtype] = None) -> np.ndarray:
```
Comment (jreback) [inline pandas/core/arrays/categorical.py:1300]: these should be NpDtype (this is a numpy protocol)

File: pandas/core/arrays/datetimelike.py [hunk 3]
```diff
@@ -265,7 +265,7 @@ def _formatter(self, boxed=False):
     # ----------------------------------------------------------------
     # Array-Like / EA-Interface Methods
 
-    def __array__(self, dtype=None) -> np.ndarray:
+    def __array__(self, dtype: Optional[Dtype] = None) -> np.ndarray:
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py:268]: same

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm ping on green

**Comment by jreback** [comment]
thanks @avinashpancham

**Pull request merged**